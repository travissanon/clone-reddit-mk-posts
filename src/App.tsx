import React, { useState, useEffect } from "react";
import getRedditPosts from "./api/getRedditPosts";
import Posts from "./client/components/Posts/index";
import Post from "./client/components/Post/index";
// import { IPost } from "./types/IPost";

export interface IPost {
	readonly id: string;
	readonly title: string;
	readonly thumbnail: string;
	readonly author: string;
}

const App: React.FC = () => {
	const [redditPosts, setRedditPosts] = useState<Array<IPost>>([]);
	const [favoritePosts, setFavoritePosts] = useState<any>(new Set());

	useEffect(() => {
		getRedditPosts().then((data) => setRedditPosts(data));
	}, []);

	const sortPosts = () => {
		setRedditPosts((prevState) => {
			const newState = [...prevState];
			newState.sort((a: any, b: any) => {
				if (a.title < b.title) {
					return -1;
				} else if (a.title > b.title) {
					return 1;
				} else {
					return 0;
				}
			});

			return newState;
		});
	};

	const toggleFavoritePost = (postId: string) => {
		setFavoritePosts((prevState: any) => {
			// spread operator is not working for Sets in Typescript for some reason.
			const newState = new Set(Array.from(prevState));

			if (newState.has(postId)) {
				newState.delete(postId);
			} else {
				newState.add(postId);
			}

			return newState;
		});
	};

	return (
		<div className="reddit-posts">
			<div className="reddit-posts__filters" onClick={() => sortPosts()}>
				Sort Posts
			</div>
			<Posts
				data={redditPosts}
				favoritePosts={favoritePosts}
				handleFavoriteClick={toggleFavoritePost}
			/>
		</div>
	);
};

export default App;
