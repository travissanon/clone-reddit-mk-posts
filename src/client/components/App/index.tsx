import React, { useState, useEffect, SetStateAction } from "react";
import getRedditPosts from "@api/getRedditPosts";
import Posts from "@components/Posts";
// import { IPost } from "./types/IPost";

export interface IPost {
	readonly id: string;
	readonly title: string;
	readonly thumbnail: string;
	readonly author: string;
}

const App: React.FC = () => {
	const [redditPosts, setRedditPosts] = useState<Array<IPost>>([]);
	const [favoritePosts, setFavoritePosts] = useState<Set<string>>(new Set());

	useEffect(() => {
		getRedditPosts().then((data) => setRedditPosts(data));
	}, []);

	const sortPosts = (): void => {
		setRedditPosts((prevState): IPost[] => {
			const newState: IPost[] = [...prevState];
			newState.sort((a: IPost, b: IPost): number => {
				if (a.title < b.title) {
					return -1;
				} else if (a.title > b.title) {
					return 1;
				} else {
					return 0;
				}
			});

			return newState;
		});
	};

	const toggleFavoritePost = (postId: string): void => {
		setFavoritePosts(
			(prevState: Set<string>): Set<string> => {
				// spread operator is not working for Sets in Typescript for some reason.
				const newState: Set<string> = new Set(Array.from(prevState));

				if (newState.has(postId)) {
					newState.delete(postId);
				} else {
					newState.add(postId);
				}

				return newState;
			}
		);
	};

	return (
		<div className="reddit-posts">
			<button
				className="reddit-posts__filters"
				onClick={(): void => sortPosts()}
			>
				Sort Posts
			</button>
			<Posts
				data={redditPosts}
				favoritePosts={favoritePosts}
				handleFavoriteClick={toggleFavoritePost}
			/>
		</div>
	);
};

export default App;
