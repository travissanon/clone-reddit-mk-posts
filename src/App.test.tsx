import * as React from "react";
import * as renderer from "react-test-renderer";

import App from "./App";

test("renders content to the page", () => {
	const tree = renderer.create(<App />).toJSON();
	expect(tree).toMatchSnapshot();
});
