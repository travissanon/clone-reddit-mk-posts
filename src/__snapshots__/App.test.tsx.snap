// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders content to the page 1`] = `
<div
  className="reddit-posts"
>
  <button
    className="reddit-posts__filters"
    onClick={[Function]}
  >
    Sort Posts
  </button>
  <div
    className="reddit-posts"
  >
    <div
      className="reddit-posts__container"
    />
  </div>
</div>
`;
